objects/src/assets/mod.rs

```diff
@@ -22,18 +22,20 @@ use core::{fmt, ops::Deref};
 /// The methodology for constructing fungible and non-fungible assets is described below.
 ///
 /// # Fungible assets
-/// The first 3 elements of a fungible asset are set to the ID of the faucet which issued the
-/// asset. This guarantees the properties described above (the first bit is ONE, and the 32 least
-/// significant bits of the 3rd element are all ZEROs).
+/// The fourth element of a fungible asset is set to the ID of the faucet which issued the asset.
+/// This guarantees the properties described above (the first bit is ONE, and the 32 least
+/// significant bits are all ZEROs).
 ///
-/// The last element is set to the amount of the asset. This amount cannot be greater than
+/// The third element is set to the amount of the asset. This amount cannot be greater than
 /// 2^63 - 1 and thus requires 63-bits to store.
 ///
+/// The first and second elements are set to ZEROs and are not used.
+///
 /// # Non-fungible assets
 /// The 4 elements of non-fungible assets are computed as follows:
 /// - First the asset data is hashed. This compresses an asset of an arbitrary length to 4 field
-///   elements.
-/// - The result of step 1 is hashed together with the ID of the faucet which issued the asset.
+///   elements: [d0, d1, d2, d3].
+/// - d3 is then replaced with the ID of the faucet which issues the asset: [d0, d1, d2, account_id].
```

Comment:
I think the rules described here (and in the comments above) need to be modified. Overall, I'm thinking of the following scheme:

* Fungible assets:
  - Layout: `[amount, 0, 0, faucet_id]`
  - Most significant bit of `faucet_id` is `1` (enforced by account ID rules).
* Non-fungible assets:
  - Layout `[h0, faucet_id, h2, d3]`
  - `h0` and `h1` come directly from `data_hash`.
  - `d3` is computed by taking `h3` from data hash and setting its most significant bit to `0`.

With this scheme, to determine if an asset is a fungible or non-fungible asset we can just look at `asset[1]`. If it is `0` we know it is a fungible asset, if it is non-zero it is a non-fungible asset (because `faucet_id` cannot be zero).